sound/soc/sof/intel/hda-loader-skl.c

```diff
@@ -0,0 +1,519 @@
+}
+
+static int
+cl_skl_cldma_copy_to_buf(struct snd_sof_dev *sdev, const void *bin,
+			 u32 total_size, u32 bufsize)
+{
+	unsigned int bytes_left = total_size;
+	const u8 *curr_pos = (u8 *)bin;
+
+	if (total_size <= 0)
+		return -EINVAL;
+
+	while (bytes_left > 0) {
+
+		if (bytes_left > bufsize) {
+
+			dev_dbg(sdev->dev, "cldma copy 0x%x bytes\n",
+				bufsize);
+
+			cl_skl_cldma_fill_buffer(sdev, bufsize,
+						 curr_pos, true, true);
+
+			bytes_left -= bufsize;
+			curr_pos += bufsize;
+		} else {
+
+			dev_dbg(sdev->dev, "cldma copy 0x%x bytes\n",
+				bytes_left);
+
+			cl_skl_cldma_set_intr(sdev, false);
+			cl_skl_cldma_fill_buffer(sdev, bytes_left,
+						 curr_pos, false, false);
+			return 0;
+		}
+	}
+
+	return bytes_left;
+}
+
+static int cl_copy_fw_skl(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_pdata *plat_data = sdev->pdata;
+	const struct firmware *fw =  plat_data->fw;
+	unsigned int bufsize = HDA_SKL_CLDMA_MAX_BUFFER_SIZE;
+	unsigned int status;
+	int ret = 0;
+
+	dev_dbg(sdev->dev, "firmware size: 0x%zx buffer size 0x%x\n", fw->size,
+		bufsize);
+
+	ret = cl_skl_cldma_copy_to_buf(sdev, fw->data, fw->size, bufsize);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: fw copy failed %d\n", ret);
+		return ret;
+	}
+
+	ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR,
+					HDA_ADSP_FW_STATUS_SKL, status,
+					((status & SKL_DSP_ROM_SKL_STS_MASK)
+						== SKL_DSP_ROM_BASEFW_ENTERED),
+					HDA_DSP_REG_POLL_INTERVAL_US,
+					HDA_DSP_BASEFW_TIMEOUT_US);
+	if (ret < 0)
+		dev_err(sdev->dev, "error: firmware loading failed!");
+
+	cl_skl_cldma_stream_run(sdev, false);
+	cl_cleanup_skl(sdev);
+
+	return ret;
+}
+
+int hda_dsp_cl_boot_firmware_skl(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	ret = cl_dsp_init_skl(sdev);
+
+	/* retry enabling core and ROM load. seemed to help */
+	if (ret < 0) {
+		ret = cl_dsp_init_skl(sdev);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: Error code=0x%x: FW status=0x%x\n",
+				snd_sof_dsp_read(sdev, HDA_DSP_BAR,
+						 HDA_ADSP_ERROR_CODE_SKL),
+				snd_sof_dsp_read(sdev, HDA_DSP_BAR,
+						 HDA_ADSP_FW_STATUS_SKL));
+			dev_err(sdev->dev, "error: Core En/ROM load fail:%d\n", ret);
+			goto irq_err;
+		}
+	}
+
+	dev_dbg(sdev->dev, "ROM init successful\n");
+
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
+	sdev->boot_complete = false;
+
+	/*
+	 * at this point DSP ROM has been initialized and should
+	 * be ready for code loading and firmware boot
+	 */
+	ret = cl_copy_fw_skl(sdev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: load firmware failed : %d\n", ret);
+		goto irq_err;
+	}
+
+	dev_dbg(sdev->dev, "Firmware download successful, booting...\n");
+	return ret;
+
+irq_err:
+	hda_dsp_dump_skl(sdev, SOF_DBG_REGS | SOF_DBG_PCI | SOF_DBG_MBOX);
+
+	/* power down DSP */
+	hda_dsp_core_reset_power_down(sdev, HDA_DSP_CORE_MASK(0));
+	cl_cleanup_skl(sdev);
+
+	dev_err(sdev->dev, "error: load fw failed err: %d\n", ret);
+	return ret;
```

Comment:
@plbossart the hda_dsp_cl_boot_firmware_skl() needs to be aligned with hda_dsp_cl_boot_firmware() to fix the retries and error handling properly.